egit 연동 및 사용을 위한 GithubMarkdownLoader 클래스 구현 및 테스트 코드 작성 완료

dadadamarine opened the pull request:
* [x] EgitUtil 테스트코드 작성
* [x] EgitUtil 구현
* resolved : #8

dadadamarine committed:
add : EgitUtil 예시md 파일 가져오기 (#8)

dadadamarine committed:
add : EgitUtil 특정url의 파일 가져오기 (#8)

dadadamarine committed:
add : EgitUtil 특정브런치의 파일 가져오기 테스트 완료 (#8)

dadadamarine committed:
add : EgitUtil 구현완료 (#8)

dadadamarine committed:
chore

hyukjin-lee commented on src/main/java/com/econo/hackday/contentsproxyblog/utils/EgitUtil.java:
```diff
@@ -1,9 +1,39 @@
+import java.io.IOException;
+import java.util.List;
 
 public class EgitUtil {
```
GithubMarkdownLoader 가 가독성 측면에서 더 낫지 않을까요

hyukjin-lee commented on src/main/java/com/econo/hackday/contentsproxyblog/utils/EgitUtil.java:
```diff
@@ -1,9 +1,39 @@
 public class EgitUtil {
 
-    public static RepositoryService repositoryService = new RepositoryService();
+    public static ContentsService contentsService = new ContentsService();
```
ContentsService는 threadSafe 한가요? Util 같은 경우에는 내부에 필드가 없고 메소드를 통해 인풋 아웃풋만 하는 경우에 주로 유틸로 선언하는데 이 ContentsService가 어떤녀석인지 좀 알아봐야할 필요가 있어보이네요. 그리고 왜 public인지 모르겠습니다.

dadadamarine replied:
- ThreadSafe의 개념에 대한 공부 : https://kldp.org/node/36904
- ContentsService 파일 : https://github.com/eclipse/egit-github/blob/master/org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/service/ContentsService.java

ContentService 클래스에서 자체적으로 전역변수를 사용하지 않고 ThreadSafe하게 동작하기때문에 
Util에서 전역변수로 사용해도 될것 같습니다.
하지만 다른사람도 똑같은 걱정을 할 수도 있고, 위험성이 말씀하신대로 존재하기때문에 ContentService를 메소드 내에서 지역변수로 생성해서 사용하는것이 좋을것 같습니다.

hyukjin-lee replied:
ThreadSafe 하다면 전역으로 빼는게 중복코드 제거나 효율성 측면에서 좋을 것 같습니다

hyukjin-lee commented on src/main/java/com/econo/hackday/contentsproxyblog/utils/EgitUtil.java:
```diff
@@ -1,9 +1,39 @@
+        }
+        String content = contents.get(0).getContent();
+
+        return new String(EncodingUtils.fromBase64(content));
```
여기 코드 쓸데없이 긴거 같습니다. 줄일 수 있어보입니다.

hyukjin-lee commented on src/main/java/com/econo/hackday/contentsproxyblog/utils/EgitUtil.java:
```diff
@@ -1,9 +1,39 @@
+        return new String(EncodingUtils.fromBase64(content));
+    }
+
+    public static RepositoryId createRepositoryId(String uri) {
```
createFromUrl 함수를 갖다 쓰면 되지 않나요?

hyukjin-lee commented on src/main/java/com/econo/hackday/contentsproxyblog/utils/EgitUtil.java:
```diff
@@ -1,9 +1,39 @@
 
+    public static String getFilePath(String uri) {
+        return uri.split("/blob")[1];
+    }
```
blob 기준으로 왜 나눠야하는건가요??

dadadamarine replied:
/blob/master 기준으로 수정하였습니다.

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -1,13 +1,16 @@
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 @RunWith(SpringRunner.class)
```
SpringRunner는 왜 붙인건가요?

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
     }
```
이건 Before 어노테이션에서 초기화하는게 맞습니다.

hyukjin-lee replied:
그리고 유틸클래스를 왜 초기화 하는건지 모르겠습니다.

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
+        ContentsService contentsService = new ContentsService();
+        String content = contentsService.getReadme(repositoryId).getContent();
+
+        assertThat(new String(EncodingUtils.fromBase64(content), "UTF-8")).isEqualTo("a\n");
```
구현된 유틸의 함수를 테스트하는게 아니라 아예 유틸을 여기에 구현한 느낌입니다.

dadadamarine replied:
실제로 위의 코드들은 메서드 테스트를 위한 테스트 코드가 아닌 처음 egit 사용방법에 대한 제가 개인적으로 테스트한 코드들 입니다. 

테스트도 공용 코드에 해당된다는 사실을 잠시 망각했습니다. pr을 날릴때는 개인코드는 삭제하도록 하겠습니다.

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
+                "- markdown -> html 렌더링\n" +
+                "    - 서버에서 할 경우 : [flexmark-java](https://github.com/vsch/flexmark-java)\n" +
+                "    - 클라이언트에서 할 경우 : [markdown-it](https://github.com/markdown-it/markdown-it)\n" +
+                "- DB : H2db\n\n");
```
여기도 마찬가지

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
+                "- markdown -> html 렌더링\n" +
+                "    - 서버에서 할 경우 : [flexmark-java](https://github.com/vsch/flexmark-java)\n" +
+                "    - 클라이언트에서 할 경우 : [markdown-it](https://github.com/markdown-it/markdown-it)\n" +
+                "- DB : H2db\n\n");
```
여기도

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
+    }
+
+    @Test
+    public void 특정_브런치_파일_가져오기_테스트() throws IOException {
```
브랜치입니다

hyukjin-lee commented on src/test/java/com/econo/hackday/contentsproxyblog/utils/EgitUtilTest.java:
```diff
@@ -18,11 +21,80 @@ public EgitUtilTest() {
+    }
+
+    @Test
+    public void 주어진url의파일가져오기테스트() throws IOException {
```
보통 한글로 쓰면 스네이크 케이스 적용해서 주어진_URL_파일_가져오기_테스트 라고 적습니다

esp2ar0 commented on src/main/java/com/econo/hackday/contentsproxyblog/exception/FileNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+package com.econo.hackday.contentsproxyblog.exception;
+
+public class FileNotFoundException extends RuntimeException {
```
FileNotFoundException은 제공되는 예외입니다

> https://docs.oracle.com/javase/7/docs/api/java/io/FileNotFoundException.html

Runtime으로 처리해주기 위해 커스텀하게 만든 건가요?
그렇다면 이름이 같으니 변경해주는게 좋을듯 싶습니다

dadadamarine replied:
실제로 커스텀 예외를 적용할 부분에 FileNotFoundException을 사용하였더니, 제공되는 예외를 참조하고있었습니다. 예외명 수정하였습니다.

hyukjin-lee commented:
PR 이름도 본인이 정한 클래스 네임을 걸어두고 그걸 구현완료 했다고하니 뭘 PR 한건지 알 수가 없습니다

dadadamarine committed:
fix :
1. util의 클래스명 변경
2. 이미 있는 예외와 이름이 중복되어 잘못 참조되는 문제 해결
3. createRepositoryId 메서드 사용 추가
(#8)

dadadamarine committed:
chore

dadadamarine committed:
chore

dadadamarine committed:
fix : GithubMarkdownLoader테스트 수정
1. 스프링과 함께 테스트하던 불필요한 과정 생략
2. 레거시 코드 제거
(#8)
